src/merkle/mmr/full.rs

```diff
@@ -0,0 +1,232 @@
+        self.forest
+    }
+
+    /// Given a leaf position, returns the merkle path to its corresponding peak. If the position
+    /// is greater-or-equal than the tree size None is returned.
+    ///
+    /// Note: The leaf position is the 0-indexed number corresponding to the order the leaves were
+    /// added, this corresponds to the MMR size _prior_ to adding the element. So the 1st alement
+    /// has position 0, the second position 1, and so on.
+    pub fn merkle_path(&self, pos: usize) -> Option<Vec<PathElement<Word>>> {
+        // find the target tree responsible for the MMR position
+        let tree_bit = leaf_to_corresponding_tree(pos, self.forest)?;
+        let forest_target = 1usize << tree_bit;
+
+        // isolate the trees before the target
+        let forest_before = self.forest & high_bitmask(tree_bit + 1);
+        let index_offset = nodes_in_forest(forest_before);
+
+        // update the position from global to the target tree
+        let mut index = nodes_in_forest(forest_target) - 1;
+        let pos = pos - forest_before;
+
+        // collect the Merkle path
+        let tree_depth = (tree_bit + 1) as usize;
+        let mut result = Vec::with_capacity(tree_depth);
+        let mut tree_depth = forest_target >> 1; // the root is not included in the path
+        while tree_depth > 0 {
+            let bit = pos & tree_depth;
+            let right_offset = index - 1;
+            let left_offset = right_offset - nodes_in_forest(tree_depth);
+
+            // Elements to the right have a higher position because they were
+            // added later. Therefore when the bit is true the node's path is
+            // to the right, and its sibling to the left.
+            let sibling = if bit != 0 {
+                index = right_offset;
+                PathElement::LeftSibling(self.nodes[index_offset + left_offset])
+            } else {
+                index = left_offset;
+                PathElement::RightSibling(self.nodes[index_offset + right_offset])
+            };
+
+            tree_depth >>= 1;
+            result.push(sibling);
+        }
+
+        Some(result)
+    }
+
+    // FUNCTIONALITY
+    // ============================================================================================
+
+    /// Adds a new element to the MMR.
+    pub fn add(&mut self, el: Word) {
+        // Note: every node is also a tree of size 1, adding an element to the forest creates a new
+        // rooted-tree of size 1. This may temporarily break the invariant that every tree in the
+        // forest has different sizes, the loop below will eagerly merge trees of same size and
+        // restore the invariant.
+        self.nodes.push(el);
+
+        let mut left_offset = self.nodes.len().saturating_sub(2);
+        let mut right = el;
+        let mut left_tree = 1;
+        while self.forest & left_tree != 0 {
+            right = *Rpo256::hash_elements(&[self.nodes[left_offset], right].concat());
```

Comment:
Why not use `Rpo256::merge()` method here?

Reply:
They are both equivalent, no? to me the top version reads slightly better:

```rust
*Rpo256::hash_elements(&[self.nodes[left_offset], right].concat());
*Rpo256::merge(&[self.nodes[left_offset].into(), right.into()])
```

but I suppose either way works.

Reply:
Yeah, I'm fine either way as well. The only reason to prefer `merge` is that it explicitly for hashing 2 digest into 1 - so, may potentially capture the intent better.